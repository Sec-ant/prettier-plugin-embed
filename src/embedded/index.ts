export * from "./register.js";
export type {
  EmbeddedDefaultComment,
  EmbeddedDefaultTag,
  EmbeddedEmbedders,
  EmbeddedLanguage,
  EmbeddedOptions,
  EmbeddedParsers,
  EmbeddedComment,
  EmbeddedTag,
  PluginEmbedOptions,
} from "./types.js";
export {
  makeIdentifiersOptionName,
  makeCommentsOptionName,
  makeTagsOptionName,
  type AutocompleteStringList,
} from "./utils.js";
