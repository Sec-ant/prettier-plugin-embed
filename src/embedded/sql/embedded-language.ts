export const embeddedLanguage = "embeddedSql";

declare module "../types.js" {
  interface EmbeddedLanguagesHolder {
    [embeddedLanguage]: void;
  }
}
