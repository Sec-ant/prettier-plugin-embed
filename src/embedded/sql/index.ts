export * from "./name.js";
export * from "./embedder.js";
export * from "./options.js";
