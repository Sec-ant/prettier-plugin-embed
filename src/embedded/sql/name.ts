export const name = "embeddedSql";

declare module "../types.js" {
  interface EmbeddedNamesHolder {
    [name]: void;
  }
}
