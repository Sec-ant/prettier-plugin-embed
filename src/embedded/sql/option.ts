import { CoreCategoryType, SupportOption } from "prettier";
export const embeddedOption: SupportOption = {
  category: "Global" satisfies CoreCategoryType,
  type: "string",
  array: true,
  default: [{ value: ["sql"] }],
  description:
    "Config embedded SQL formatting. This requires prettier-plugin-sql",
};
