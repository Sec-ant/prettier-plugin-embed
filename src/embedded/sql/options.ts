import type { ChoiceSupportOption, SupportOptions } from "prettier";
import {
  type AutocompleteStringList,
  type StringListToInterfaceKey,
  fallbackIndicator,
  makeCommentsOptionName,
  makeIdentifiersOptionName,
  makeParserOptionName,
  makePluginOptionName,
  makeTagsOptionName,
} from "../utils.js";
import { language } from "./language.js";

const DEFAULT_COMMENTS_OR_TAGS = ["sql"] as const;

const DEFAULT_COMMENTS = DEFAULT_COMMENTS_OR_TAGS;
type Comments = AutocompleteStringList<(typeof DEFAULT_COMMENTS)[number]>;
type DefaultCommentsHolder = StringListToInterfaceKey<typeof DEFAULT_COMMENTS>;

const DEFAULT_TAGS = DEFAULT_COMMENTS_OR_TAGS;
type Tags = AutocompleteStringList<(typeof DEFAULT_TAGS)[number]>;
type DefaultTagsHolder = StringListToInterfaceKey<typeof DEFAULT_TAGS>;

const SQL_PLUGINS = ["prettier-plugin-sql", "prettier-plugin-sql-cst"] as const;
type SqlPlugin = (typeof SQL_PLUGINS)[number];

const SQL_CST_PARSERS = [
  "sqlite",
  "bigquery",
  "mysql",
  "mariadb",
  "postgresql",
] as const;
type SqlCstParser = (typeof SQL_CST_PARSERS)[number];

const EMBEDDED_LANGUAGE_IDENTIFIERS = makeIdentifiersOptionName(language);

const EMBEDDED_LANGUAGE_COMMENTS = makeCommentsOptionName(language);
const EMBEDDED_LANGUAGE_TAGS = makeTagsOptionName(language);
const EMBEDDED_LANGUAGE_PLUGIN = makePluginOptionName(language);
const EMBEDDED_LANGUAGE_PARSER = makeParserOptionName(language);

export const options = {
  [EMBEDDED_LANGUAGE_IDENTIFIERS]: {
    category: "Embed",
    type: "string",
    array: true,
    default: [{ value: [...DEFAULT_COMMENTS_OR_TAGS] }],
    description:
      "Tag or comment identifiers that make their subsequent template literals be identified as embedded SQL language. This option requires the `prettier-plugin-sql` plugin or the `prettier-plugin-sql-cst` plugin.",
    deprecated: `Please use \`${EMBEDDED_LANGUAGE_COMMENTS}\` or \`${EMBEDDED_LANGUAGE_TAGS}\`.`,
  },
  [EMBEDDED_LANGUAGE_COMMENTS]: {
    category: "Embed",
    type: "string",
    array: true,
    default: [{ value: [fallbackIndicator] }],
    description:
      "Block comments that make their subsequent template literals be identified as embedded SQL language. This option requires the `prettier-plugin-sql` plugin or the `prettier-plugin-sql-cst` plugin.",
  },
  [EMBEDDED_LANGUAGE_TAGS]: {
    category: "Embed",
    type: "string",
    array: true,
    default: [{ value: [fallbackIndicator] }],
    description:
      "Tags that make their subsequent template literals be identified as embedded SQL language. This option requires the `prettier-plugin-sql` plugin or the `prettier-plugin-sql-cst` plugin.",
  },
  [EMBEDDED_LANGUAGE_PLUGIN]: {
    category: "Embed",
    type: "choice",
    default: "prettier-plugin-sql",
    description:
      "The plugin used to format the embedded SQL language. This option requires the `prettier-plugin-sql` plugin or the `prettier-plugin-sql-cst` plugin.",
    choices: SQL_PLUGINS.map((plugin) => ({
      value: plugin,
      description: `Use the "${plugin}" plugin.`,
    })),
  } satisfies ChoiceSupportOption<SqlPlugin>,
  [EMBEDDED_LANGUAGE_PARSER]: {
    category: "Embed",
    type: "choice",
    default: "sqlite",
    description:
      "Specify the embedded SQL language parser. This option is only needed with the `prettier-plugin-sql-cst` plugin.",
    choices: SQL_CST_PARSERS.map((parser) => ({
      value: parser,
      description: `Use the "${parser}" parser.`,
    })),
  } satisfies ChoiceSupportOption<SqlCstParser>,
} as const satisfies SupportOptions;

type Options = typeof options;

declare module "../types.js" {
  interface EmbeddedOptions extends Options {}
  interface EmbeddedDefaultCommentsHolder extends DefaultCommentsHolder {}
  interface EmbeddedDefaultTagsHolder extends DefaultTagsHolder {}
  interface PluginEmbedOptions {
    /**
     * @deprecated Please use `embeddedSqlComments` or `embeddedSqlTags`.
     */
    [EMBEDDED_LANGUAGE_IDENTIFIERS]?: (Comments[number] | Tags[number])[];
    [EMBEDDED_LANGUAGE_COMMENTS]?: Comments;
    [EMBEDDED_LANGUAGE_TAGS]?: Tags;
    [EMBEDDED_LANGUAGE_PLUGIN]?: SqlPlugin;
    [EMBEDDED_LANGUAGE_PARSER]?: SqlCstParser;
  }
}
