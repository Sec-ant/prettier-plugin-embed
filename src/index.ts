export { options } from "./options.js";
export { parsers } from "./parsers.js";
export { printers } from "./printers.js";
export type { PluginEmbedOptions } from "./embedded/index.js";
export type { EmbeddedOverride } from "./types.js";

export type {
  /**
   * @deprecated Renamed to `PluginEmbedOptions`
   */
  PluginEmbedOptions as PrettierPluginEmbedOptions,
} from "./embedded/index.js";
