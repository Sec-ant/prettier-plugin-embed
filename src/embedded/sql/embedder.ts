import type { Options } from "prettier";
import type { SqlBaseOptions } from "prettier-plugin-sql";
import { builders, utils } from "prettier/doc";
import type { Embedder } from "../../types.js";
import {
  preparePlaceholder,
  printTemplateExpressions,
  throwIfPluginIsNotFound,
} from "../utils.js";
import { embeddedLanguage } from "./embedded-language.js";
import {
  NODE_SQL_PARSER_DATABASES,
  SQL_FORMATTER_LANGUAGES,
  type NodeSqlParserDataBase,
  type SqlFormatterLanguage,
} from "./options.js";

const { hardline, group, line, softline, indent } = builders;
const { mapDoc } = utils;

export const embedder: Embedder<Options> = async (
  textToDoc,
  print,
  path,
  options,
  { identifier, identifiers, embeddedOverrideOptions },
) => {
  throwIfPluginIsNotFound("prettier-plugin-sql", options, identifier);

  options = {
    ...options,
    ...embeddedOverrideOptions,
  };

  const { node } = path;

  const { createPlaceholder, placeholderRegex } = preparePlaceholder();

  const text = node.quasis
    .map((quasi, index, { length }) =>
      index === length - 1
        ? quasi.value.cooked
        : quasi.value.cooked + createPlaceholder(index),
    )
    .join("");

  const leadingWhitespaces = text.match(/^\s+/)?.[0] ?? "";
  const trailingWhitespaces = text.match(/\s+$/)?.[0] ?? "";

  const expressionDocs = printTemplateExpressions(path, print);

  const optionsOverride = getOptionsOverride(options, identifier, identifiers);
  if (typeof optionsOverride === "undefined") {
    throw new SyntaxError(`Unrecognized identifier: ${identifier}`);
  }

  const doc = await textToDoc(text, {
    ...options,
    parser: "sql",
    ...optionsOverride,
    ...embeddedOverrideOptions,
  });

  const contentDoc = mapDoc(doc, (doc) => {
    if (typeof doc !== "string") {
      return doc;
    }
    const parts = [];
    const components = doc.split(placeholderRegex);
    for (let i = 0; i < components.length; i++) {
      let component = components[i];
      if (i % 2 == 0) {
        if (!component) {
          continue;
        }
        component = component.replaceAll(/([\\`]|\${)/g, "\\$1");
        component
          .split(/(\n)/)
          .forEach((c) => (c === "\n" ? parts.push(hardline) : parts.push(c)));
      } else {
        const placeholderIndex = Number(component);
        parts.push(expressionDocs[placeholderIndex]);
      }
    }
    return parts;
  });

  if (options.preserveEmbeddedExteriorWhitespaces?.includes(identifier)) {
    // TODO: should we label the doc with { hug: false } ?
    // https://github.com/prettier/prettier/blob/5cfb76ee50cf286cab267cf3cb7a26e749c995f7/src/language-js/embed/html.js#L88
    return group([
      "`",
      leadingWhitespaces,
      options.noEmbeddedMultiLineIndentation?.includes(identifier)
        ? [group(contentDoc)]
        : indent([group(contentDoc)]),
      trailingWhitespaces,
      "`",
    ]);
  }

  const leadingLineBreak = leadingWhitespaces.length ? line : softline;
  const trailingLineBreak = trailingWhitespaces.length ? line : softline;

  return group([
    "`",
    options.noEmbeddedMultiLineIndentation?.includes(identifier)
      ? [leadingLineBreak, group(contentDoc)]
      : indent([leadingLineBreak, group(contentDoc)]),
    trailingLineBreak,
    "`",
  ]);
};

type Formatter = Exclude<Options["formatter"], undefined>;

const testFormatterIdentifiers: Record<
  Formatter,
  (identifier: string) => boolean
> = {
  "sql-formatter": (identifier: string) =>
    SQL_FORMATTER_LANGUAGES.includes(identifier as SqlFormatterLanguage),
  "node-sql-parser": (identifier: string) =>
    NODE_SQL_PARSER_DATABASES.includes(identifier as NodeSqlParserDataBase),
  // TODO: This seems not implemented yet
  "sql-cst": (identifier: string) =>
    SQL_FORMATTER_LANGUAGES.includes(identifier as SqlFormatterLanguage),
};

function getOptionsOverrideByFormatters(
  identifier: string,
  formatters: Formatter[],
): SqlBaseOptions | undefined {
  for (const formatter of formatters) {
    if (testFormatterIdentifiers[formatter](identifier)) {
      switch (formatter) {
        case "sql-formatter":
          return {
            formatter,
            language: identifier,
          };
        case "node-sql-parser":
          return {
            formatter,
            database: identifier,
          };
      }
    }
  }
}

function getOptionsOverrideByFormattersWithFallback(
  identifier: string,
  identifiers: string[],
  formatters: Formatter[],
): SqlBaseOptions | undefined {
  let optionsOverride = getOptionsOverrideByFormatters(identifier, formatters);
  if (optionsOverride) {
    return optionsOverride;
  }
  const index = identifiers.indexOf(identifier);
  for (let i = index - 1; i >= 0; --i) {
    optionsOverride = getOptionsOverrideByFormatters(
      identifiers[i],
      formatters,
    );
    if (optionsOverride) {
      return optionsOverride;
    }
  }
}

function getOptionsOverride(
  options: Options,
  identifier: string,
  identifiers: string[],
): SqlBaseOptions | undefined {
  const formatter = options.formatter ?? "sql-formatter";
  switch (formatter) {
    case "sql-formatter":
      return getOptionsOverrideByFormattersWithFallback(
        identifier,
        identifiers,
        ["sql-formatter", "node-sql-parser"],
      );
    case "node-sql-parser":
      return getOptionsOverrideByFormattersWithFallback(
        identifier,
        identifiers,
        ["node-sql-parser", "sql-formatter"],
      );
    case "sql-cst":
      // TODO: This seems not implemented yet
      return getOptionsOverrideByFormattersWithFallback(
        identifier,
        identifiers,
        ["sql-formatter", "node-sql-parser"],
      );
    default:
      // guard
      formatter satisfies never;
      // by falling back to the sql-formatter
      // we use a rather forgiving strategy
      // to parse the options
      return getOptionsOverrideByFormattersWithFallback(
        identifier,
        identifiers,
        ["sql-formatter", "node-sql-parser"],
      );
  }
}

declare module "../types.js" {
  interface EmbeddedEmbedders {
    [embeddedLanguage]: typeof embedder;
  }
}
