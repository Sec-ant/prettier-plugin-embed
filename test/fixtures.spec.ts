import fs from "node:fs/promises";
import path from "node:path";

import prettier from "prettier";
import sql from "prettier-plugin-sql";
import * as embed from "../src";

import { describe, expect, it } from "vitest";

describe("fixtures", () => {
  it("should work", async () => {
    for (const file of await fs.readdir("test/fixtures")) {
      const filepath = path.resolve("test/fixtures", file);
      const code = await fs.readFile(filepath, "utf-8");
      const formatted = await prettier.format(code, {
        filepath,
        plugins: [embed, sql],
        embeddedSqlTags: ["sql.type"],
      });
      expect(formatted).toMatchSnapshot();
    }
  });
});
