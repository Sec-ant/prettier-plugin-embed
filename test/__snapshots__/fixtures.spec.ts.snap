// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > should work 1`] = `
"sql.type(SESSION)\`
  INSERT INTO
    sessions (account_id, expires_at)
  VALUES
    (
      \${accountId},
      \${expiresAt.toJSON()}
    ) RETURNING id,
    created_at AS "createdAt",
    updated_at AS "updatedAt",
    expires_at AS "expiresAt",
    account_id AS "accountId"
\`;
"
`;
